### PR TITLE
implementation of exit, shutdown

### PR DESCRIPTION
**Pull request by lachhebo**
I tried to implement the exit and shutdown of LSP without success, 

When receiving an exit notification, the server should exit with a 0 code.
When receiving an shutdown request, the server should send a message null.

In this situation, the sever won't start.

### CODE REVIEW

**Review by ejgallego — COMMENTED** [review]
So what is missing? Apart from calling `exit 0` on the exit case, the rest looks good to me.

**Comment by lachhebo** [comment]
the server is unable to respond to the initialize request with this implementation on my machine

**Comment by lachhebo** [comment]
Sorry, it was just a small error, it is able ta start the server, but not yet to shut it.

File: lp-lsp/lp_lsp.ml [hunk 1]
```diff
@@ -63,11 +69,12 @@ let dispatch_method ofmt dict =
     do_change ofmt params
   | "textDocument/didClose" ->
     do_close ofmt params
+  | "shutdown" ->
+    do_shutdown ofmt
+  | "exit" -> exit 0;
```
Comment (ejgallego) [inline lp-lsp/lp_lsp.ml:74]: Careful, no ";" goes here!

**Comment by ejgallego** [comment]
Add the following code before `module LIO`:
```ocaml
(* Conditionals *)
let option_default x d = match x with | None -> d | Some x -> x
let odict_field name dict = option_default U.(to_option to_assoc List.(assoc name dict)) []
```
now modify the dispatch so that:
```ocaml
let dispatch_method ofmt dict =
  let params = odict_field "params" dict in
  match string_field "method" dict with
```

that will avoid the problem you were facing.

**Comment by lachhebo** [comment]
Thank you Emilio, it is working well now

**Comment by ejgallego** [comment]
Can you update the PR so I merge?

**Comment by ejgallego** [comment]
Ok, I merged this on my tree, had to rebase anyways to follow upstream, thanks!

**Pull request closed without merging**